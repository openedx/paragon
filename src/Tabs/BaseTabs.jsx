import React from 'react';
import PropTypes from 'prop-types';
import { useUncontrolled } from 'uncontrollable';
import TabContainer from 'react-bootstrap/TabContainer';
import TabContent from 'react-bootstrap/TabContent';
import TabPane from 'react-bootstrap/TabPane';
import { forEach, map } from 'react-bootstrap/ElementChildren';
import Nav, { NavLink, NavItem } from '../Nav';

function getDefaultActiveKey(children) {
  let defaultActiveKey;

  forEach(children, (child) => {
    if (defaultActiveKey == null) {
      defaultActiveKey = child.props.eventKey;
    }
  });

  return defaultActiveKey;
}

function renderTab(child) {
  const {
    title, eventKey, disabled, tabClassName, id,
  } = child.props;

  if (title == null) {
    return null;
  }

  return (
    <NavLink
      as="li"
      eventKey={eventKey}
      disabled={disabled}
      id={id}
      className={tabClassName}
    >
      {title}
    </NavLink>
  );
}

function BaseTabs(props) {
  const {
    id,
    onSelect,
    transition,
    mountOnEnter,
    unmountOnExit,
    children,
    activeKey = getDefaultActiveKey(children),
    ...controlledProps
  } = useUncontrolled(props, {
    activeKey: 'onSelect',
  });

  return (
    <TabContainer
      id={id}
      activeKey={activeKey}
      onSelect={onSelect}
      transition={transition}
      mountOnEnter={mountOnEnter}
      unmountOnExit={unmountOnExit}
    >
      <Nav {...controlledProps} role="tablist" as="ul">
        {map(children, renderTab)}
      </Nav>

      <TabContent>
        {map(children, (child) => {
          const childProps = { ...child.props };

          delete childProps.title;
          delete childProps.disabled;
          delete childProps.tabClassName;

          return <TabPane {...childProps} />;
        })}
      </TabContent>
    </TabContainer>
  );
}

BaseTabs.propTypes = {
  activeKey: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  defaultActiveKey: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onSelect: PropTypes.func,
  variant: PropTypes.string,
  transition: PropTypes.oneOfType([
    PropTypes.oneOf([false]),
    PropTypes.elementType,
  ]),
  id: PropTypes.string,
  mountOnEnter: PropTypes.bool,
  unmountOnExit: PropTypes.bool,
};

BaseTabs.defaultProps = {
  activeKey: undefined,
  defaultActiveKey: undefined,
  onSelect: undefined,
  transition: undefined,
  variant: 'tabs',
  id: undefined,
  mountOnEnter: false,
  unmountOnExit: false,
};

BaseTabs.displayName = 'Tabs';

export default BaseTabs;
