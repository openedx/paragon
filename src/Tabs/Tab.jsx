import React from 'react';
import PropTypes from 'prop-types';
import BaseTab from 'react-bootstrap/Tab';

const Tab = (props) => <BaseTab {...props} />;

Tab.propTypes = {
  /** Specifies the `Tab` navigation title. */
  title: PropTypes.node.isRequired,
  /** Specifies notification bubble content. It appears on the top right of the `Tab`. */
  notification: PropTypes.node,
  /** Specifies whether `Tab` is disabled. */
  disabled: PropTypes.bool,
  /**
   * A unique identifier for the Component, the `eventKey` makes it distinguishable
   * from others in a set. Similar to React's `key` prop, in that it only needs to be
   * unique amongst the Components siblings, not globally.
   */
  eventKey: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** Specifies class name to append to the base element. */
  tabClassName: PropTypes.string,
};

Tab.defaultProps = {
  notification: undefined,
  disabled: undefined,
  eventKey: undefined,
  tabClassName: undefined,
};

export default Tab;
