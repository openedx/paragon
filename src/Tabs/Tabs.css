/**
 * Do not edit directly
 * Generated on Wed, 01 Mar 2023 00:27:20 GMT
 */

:root {
  --pgn-size-tabs-notification-width: 1rem;
  --pgn-size-tabs-notification-height: 1rem;
  --pgn-typography-tabs-notification-font-size: var(--pgn-typography-font-size-xs);
}
