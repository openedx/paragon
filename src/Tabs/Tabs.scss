@import "variables";

.pgn__tabs {
  flex-wrap: nowrap;

  & > * {
    position: relative;
  }

  .pgn__tab-notification {
    position: absolute;
    top: 0;
    right: 0;
    min-height: $tab-notification-height;
    min-width: $tab-notification-width;
    font-size: $tab-notification-font-size;
  }
}

.pgn__tab_invisible {
  position: absolute;
  left: 0;
  pointer-events: none;
  visibility: hidden;
}

.pgn__tab_more.nav-link {
  margin-bottom: 0;
  padding: 0;
  border-top: none;
  border-bottom: none;

  .dropdown .dropdown-toggle {
    display: block;
    padding: .5rem 1rem;
    border-top-width: 4px;
    border-bottom-width: 4px;
  }

  &.active {
    .dropdown .dropdown-toggle {
      // stylelint-disable-next-line
      @extend .nav-link, .active;
    }
  }
}
