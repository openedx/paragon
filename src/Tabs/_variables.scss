$tab-notification-height:       1rem !default;
$tab-notification-width:        1rem !default;
$tab-notification-font-size:    $font-size-xs !default;
