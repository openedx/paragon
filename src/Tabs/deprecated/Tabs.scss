.nav-tabs .nav-link {
  background: transparent;
}
