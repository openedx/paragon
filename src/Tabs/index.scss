@import "variables";

.pgn__tabs {
  flex-wrap: nowrap;

  & > * {
    position: relative;
  }

  .pgn__tab-notification {
    position: absolute;
    top: 0;
    right: 0;
    min-height: $tab-notification-height;
    min-width: $tab-notification-width;
    font-size: $tab-notification-font-size;
  }

  .pgn__tab_more.nav-link {
    margin-bottom: 0;
    padding: 0;

    .dropdown .dropdown-toggle {
      display: block;
      border: none;
    }
  }

  // Nav-pills
  &.nav-pills {
    .pgn__tab_more.nav-link {
      margin: 0;

      .dropdown .dropdown-toggle {
        padding: $tab-more-link-dropdown-toggle-padding-x $tab-more-link-dropdown-toggle-padding-y;

        &:focus {
          background-color: $tab-more-link-dropdown-toggle-focus-bg;
          border-color: $tab-more-link-dropdown-toggle-focus-border-color;
          color: $tab-more-link-dropdown-toggle-focus-color;

          &.btn.btn-link {
            color: $tab-more-link-dropdown-toggle-btn-focus-color;
            border-color: $tab-more-link-dropdown-toggle-btn-focus-border-color;
          }
        }
      }

      &.active {
        .dropdown .dropdown-toggle {
          color: $tab-more-link-dropdown-toggle-active-color;
        }

        &:hover {
          .dropdown .dropdown-toggle {
            background-color: $tab-more-link-dropdown-toggle-hover-color;
          }
        }
      }
    }
  }

  // Nav inverse pills
  &.nav-inverse-pills .pgn__tab_more.nav-link {
    .dropdown .dropdown-toggle {
      padding: $tab-inverse-pills-link-dropdown-toggle-padding-x $tab-inverse-pills-link-dropdown-toggle-padding-y;
      border: none;

      &:focus {
        background-color: $tab-inverse-pills-link-dropdown-toggle-focus-bg;
        color: $tab-inverse-pills-link-dropdown-toggle-focus-color;
      }
    }

    &.active .dropdown .dropdown-toggle {
      color: $tab-inverse-pills-link-dropdown-toggle-active-color;

      &:hover {
        color: $tab-inverse-pills-link-dropdown-toggle-active-hover-color;
        background-color: $tab-inverse-pills-link-dropdown-toggle-active-hover-bg;
      }

      &:focus {
        @include nav-tabs-link-focus(
          $tab-inverse-pills-link-dropdown-toggle-focus-border-color,
          $nav-tabs-border-radius,
          $tab-inverse-pills-link-dropdown-distance
        );
      }
    }
  }

  // Nav tabs
  &.nav-tabs .pgn__tab_more.nav-link .dropdown .dropdown-toggle {
    padding: .563rem $tab-more-link-dropdown-toggle-padding-y;
  }

  // Nav inverse tabs
  &.nav-inverse-tabs .pgn__tab_more.nav-link .dropdown .dropdown-toggle {
    padding: $tab-inverse-tabs-link-dropdown-toggle-padding-x $tab-inverse-tabs-link-dropdown-toggle-padding-y;

    &:hover {
      background-color: $tab-inverse-tabs-link-dropdown-toggle-hover-bg;
    }
  }

  .pgn__tab_invisible {
    position: absolute;
    left: 0;
    pointer-events: none;
    visibility: hidden;
  }
}
