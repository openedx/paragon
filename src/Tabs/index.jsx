import React, { useEffect, useMemo, useRef } from 'react';
import BaseTabs from 'react-bootstrap/Tabs';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import TabsDeprecated from './deprecated';
import Bubble from '../Bubble';
import Dropdown from '../Dropdown';
import useIndexOfLastVisibleChild from '../hooks/useIndexOfLastVisibleChild';
import Tab from './Tab';

export const MORE_TAB_TEXT = 'More...';

function Tabs({
  children,
  className,
  moreTabText = MORE_TAB_TEXT,
  defaultActiveKey,
  activeKey,
  ...props
}) {
  const containerElementRef = useRef(null);
  const overflowElementRef = useRef(null);
  const indexOfLastVisibleChild = useIndexOfLastVisibleChild(
    containerElementRef.current?.children[0],
    overflowElementRef.current?.parentNode,
  );

  useEffect(() => {
    if (containerElementRef.current) {
      const observer = new MutationObserver((mutations => {
        mutations.forEach(mutation => {
          // React-Bootstrap attribute 'data-rb-event-key' is responsible for the tab identification
          const eventKey = mutation.target.getAttribute('data-rb-event-key');
          // React-Bootstrap attribute 'aria-selected' is responsible for selected/unselected state
          const isActive = mutation.target.getAttribute('aria-selected') === 'true';
          // datakey attribute is added manually to the dropdown
          // elements so that they correspond to the native tabs' eventKey
          const element = containerElementRef.current.querySelector(`[datakey='${eventKey}']`);
          const moreTab = containerElementRef.current.querySelector('.pgn__tab_more');
          if (isActive) {
            element?.classList.add('active');
            // Here we add active class to the 'More Tab' if element exists in the dropdown
            if (element) {
              moreTab.classList.add('active');
            } else {
              moreTab.classList.remove('active');
            }
          } else {
            element?.classList.remove('active');
          }
        });
      }));
      observer.observe(containerElementRef.current, {
        attributes: true, subtree: true, attributeFilter: ['aria-selected'],
      });
      return () => observer.disconnect();
    }
    return undefined;
  }, []);

  useEffect(() => {
    if (overflowElementRef.current?.parentNode) {
      overflowElementRef.current.parentNode.tabIndex = -1;
    }
  }, [overflowElementRef.current?.parentNode]);

  const handleDropdownTabClick = (eventKey) => {
    const hiddenTab = containerElementRef.current.querySelector(`[data-rb-event-key='${eventKey}']`);
    hiddenTab.click();
  };

  const tabsChildren = useMemo(() => {
    const indexOfOverflowStart = indexOfLastVisibleChild + 1;
    const childrenList = React.Children.map(children, (child, index) => {
      if (child?.type?.name !== 'Tab' && process.env.NODE_ENV === 'development') {
        // eslint-disable-next-line no-console
        console.error(
          `Tabs children can only be of type Tab. ${children[index]} was passed instead.`,
        );
      }
      if (!React.isValidElement(child)) {
        return child;
      }
      const {
        title, notification, tabClassName, ...rest
      } = child.props;
      let newTitle;
      if (notification) {
        newTitle = (
          <>
            {title}
            <Bubble
              variant="error"
              role="status"
              className="pgn__tab-notification"
              expandable
            >
              {notification}
            </Bubble>
          </>
        );
      } else {
        newTitle = title;
      }
      const tabClass = index > indexOfLastVisibleChild ? 'pgn__tab_invisible' : '';
      const modifiedTab = React.cloneElement(child, {
        ...rest,
        title: newTitle,
        tabClassName: classNames(tabClass, tabClassName),
      });
      return modifiedTab;
    });
    let moreTabHasNotification = false;
    const overflowChildren = childrenList.slice(indexOfOverflowStart)
      .map(overflowChild => {
        if (!moreTabHasNotification && overflowChild.props.notification) {
          moreTabHasNotification = true;
        }
        return (
          <Dropdown.Item
            key={`${overflowChild.props.eventKey}overflow`}
            onClick={() => handleDropdownTabClick(overflowChild.props.eventKey)}
            disabled={overflowChild.props.disabled}
            datakey={overflowChild.props.eventKey}
            className={classNames({
              active: overflowChild.props.eventKey === defaultActiveKey || overflowChild.props.eventKey === activeKey,
            }, 'pgn__tabs__dropdown-item')}
          >
            {overflowChild.props.title}
          </Dropdown.Item>
        );
      });

    childrenList.splice(indexOfOverflowStart, 0, (<Tab
      key="moreTabKey"
      tabClassName={classNames(!overflowChildren.length && 'pgn__tab_invisible', 'pgn__tab_more')}
      title={(
        <Dropdown ref={overflowElementRef}>
          <Dropdown.Toggle
            variant="link"
            className="nav-link"
            id="pgn__tab-toggle"
          >
            {moreTabText}
            {moreTabHasNotification && (
              <Bubble
                variant="error"
                role="status"
                className="pgn__tab-notification"
              />
            )}
          </Dropdown.Toggle>
          <Dropdown.Menu className="dropdown-menu-right">{overflowChildren}</Dropdown.Menu>
        </Dropdown>
      )}
    />
    ));
    return childrenList;
  }, [activeKey, children, defaultActiveKey, indexOfLastVisibleChild, moreTabText]);

  return (
    <div ref={containerElementRef}>
      <BaseTabs
        defaultActiveKey={defaultActiveKey}
        activeKey={activeKey}
        {...props}
        className={classNames(className, 'pgn__tabs')}
      >
        {tabsChildren}
      </BaseTabs>
    </div>
  );
}

Tabs.propTypes = {
  /** Specifies elements that is processed to create tabs. */
  children: PropTypes.node.isRequired,
  /** Specifies class name to append to the base element. */
  className: PropTypes.string,
  /** Specifies text for the 'More' tab. */
  moreTabText: PropTypes.string,
  /** Specifies default active tab (uncontrolled usage). */
  defaultActiveKey: PropTypes.string,
  /** Specifies active tab (controlled usage). */
  activeKey: PropTypes.string,
};

Tabs.defaultProps = {
  className: undefined,
  moreTabText: MORE_TAB_TEXT,
  defaultActiveKey: undefined,
  activeKey: undefined,
};

Tabs.Deprecated = TabsDeprecated;

export default Tabs;
export { Tab };
export { default as TabContainer } from 'react-bootstrap/TabContainer';
export { default as TabContent } from 'react-bootstrap/TabContent';
export { default as TabPane } from 'react-bootstrap/TabPane';
