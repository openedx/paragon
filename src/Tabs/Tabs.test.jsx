import React from 'react';
import { render, waitFor, act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';

import Tabs, { MORE_TAB_TEXT } from '.';
import Tab from './Tab';

jest.mock('../hooks/useIndexOfLastVisibleChild', () => ({
  __esModule: true,
  default: () => 0,
}));

let getAttributeValue = 'true';

const disconnect = jest.fn();
const observeMutation = jest.fn((callback) => callback([{
  target: {
    getAttribute: () => getAttributeValue,
  },
}]));
const observeResize = jest.fn();
const unobserve = jest.fn();

window.MutationObserver = jest.fn((callback) => ({
  disconnect,
  observe: () => observeMutation(callback),
  unobserve,
}));

window.ResizeObserver = jest.fn(() => ({
  disconnect,
  observe: observeResize,
  unobserve,
}));

function TabsTestComponent(props) {
  return (
    <Tabs {...props} defaultActiveKey="tab_1">
      <Tab title="Tab 1" eventKey="tab_1" notification={4} />
      <Tab title="Tab 2" eventKey="tab_2" />
      <Tab title="Tab 3" eventKey="tab_3" />
    </Tabs>
  );
}

describe('<Tabs />', () => {
  afterEach(() => {
    document.getElementsByTagName('html')[0].innerHTML = '';
    observeMutation.mockClear();
  });

  describe('correct rendering', () => {
    it('renders successfully', () => {
      const tree = renderer.create(<TabsTestComponent />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders notification property', () => {
      const { getAllByRole } = render(<TabsTestComponent />);

      const notification = getAllByRole('status');
      expect(notification.length).toBeGreaterThan(0);
    });
    it('MutationObserver is initialized', () => {
      render(<TabsTestComponent />);
      expect(observeMutation).toHaveBeenCalledTimes(1);

      observeMutation.mockClear();
      getAttributeValue = 'false';
      render(<TabsTestComponent />);
      expect(observeMutation).toHaveBeenCalledTimes(1);
    });
    it('dropdown menu is displayed', () => {
      const { getAllByRole } = render(<TabsTestComponent />);
      const dropdownMenu = getAllByRole('tabpanel', { hidden: true })[1];
      expect(dropdownMenu).toBeInTheDocument();
      expect(dropdownMenu.className).not.toContain('pgn__tab_invisible');
    });
    it('moreTabText is displayed', async () => {
      const text = 'Mehr...';
      const { rerender, getByText } = render(<TabsTestComponent />);
      const toggleButton = getByText(MORE_TAB_TEXT);
      expect(toggleButton).toBeInTheDocument();
      await act(async () => {
        await userEvent.click(toggleButton);
      });
      rerender(<TabsTestComponent moreTabText={text} />);
      expect(toggleButton.textContent).toBe(text);
    });
    it('click on the dropdown item activates tab', async () => {
      const { container, getByText, getAllByText } = render(<TabsTestComponent />);
      const toggleButton = getByText(MORE_TAB_TEXT);
      await act(async () => {
        await userEvent.click(toggleButton);
      });
      const dropdownItem = getAllByText('Tab 2');
      await act(async () => {
        await userEvent.click(dropdownItem[0], undefined, { skipPointerEventsCheck: true });
      });
      const tab = container.querySelector('[data-rb-event-key="tab_2"]');
      expect(tab.className).toContain('active');
    });
    it('select dropdown item after pressing Enter', async () => {
      const { getByText, getAllByText, getByRole } = render(<TabsTestComponent />);
      const toggleButton = getByText(MORE_TAB_TEXT);
      await act(async () => {
        await userEvent.click(toggleButton);
      });
      const dropdownItem = getAllByText('Tab 2');
      dropdownItem[0].focus();
      await act(async () => {
        await userEvent.keyboard('{enter}');
      });
      await waitFor(() => {
        const tab = getByRole('tab', { name: 'Tab 2' });
        expect(tab.className).toContain('active');
      });
    });
    it('invalid child does not render', () => {
      const { getByRole } = render(
        <Tabs>
          {[false, undefined]}
        </Tabs>,
      );
      const navElement = getByRole('tablist');
      const childElements = navElement.children;
      expect(childElements.length).toEqual(1);
    });
  });
});
