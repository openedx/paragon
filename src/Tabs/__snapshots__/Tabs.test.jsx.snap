// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> correct rendering renders successfully 1`] = `
<div>
  <nav
    className="pgn__tabs nav nav-tabs"
    onKeyDown={[Function]}
    role="tablist"
  >
    <a
      aria-selected={true}
      className="nav-item nav-link active"
      data-rb-event-key="tab_1"
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
    >
      Tab 1
      <div
        className="pgn__bubble pgn__bubble-error pgn__tab-notification expandable"
        role="status"
      >
        4
      </div>
    </a>
    <a
      className="pgn__tab_more nav-item nav-link"
      data-rb-event-key={null}
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
    >
      <div
        className="pgn__dropdown pgn__dropdown-light dropdown"
        data-testid="dropdown"
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="nav-link dropdown-toggle btn btn-link"
          disabled={false}
          id="pgn__tab-toggle"
          onClick={[Function]}
          type="button"
        >
          More...
        </button>
      </div>
    </a>
    <a
      aria-selected={false}
      className="pgn__tab_invisible nav-item nav-link"
      data-rb-event-key="tab_2"
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
    >
      Tab 2
    </a>
    <a
      aria-selected={false}
      className="pgn__tab_invisible nav-item nav-link"
      data-rb-event-key="tab_3"
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
    >
      Tab 3
    </a>
  </nav>
  <div
    className="tab-content"
  >
    <div
      aria-hidden={false}
      aria-labelledby={null}
      className="fade tab-pane active show"
      id={null}
      notification={4}
      role="tabpanel"
    />
    <div
      aria-hidden={true}
      aria-labelledby={null}
      className="fade tab-pane"
      id={null}
      role="tabpanel"
    />
    <div
      aria-hidden={true}
      aria-labelledby={null}
      className="fade tab-pane"
      id={null}
      role="tabpanel"
    />
    <div
      aria-hidden={true}
      aria-labelledby={null}
      className="fade tab-pane"
      id={null}
      role="tabpanel"
    />
  </div>
</div>
`;
